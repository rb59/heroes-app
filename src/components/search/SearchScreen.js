import React, { useMemo } from 'react';
import queryString from 'query-string';
import { useForm } from '../../hooks/useForm';
import { HeroCard } from '../heroes/HeroCard';
import { useLocation } from 'react-router-dom';
import { getHeroesByName } from '../../selectors/getHeroesByName';
// cd
export const SearchScreen = ({history}) => {

    const location = useLocation();
    const {q = ''} = queryString.parse(location.search)
    const [searchValues,handleInputChange] = useForm({
        search: q,
    });
    const {search} = searchValues;
    const heroesFiltered = useMemo(() => getHeroesByName(q), [q]) ;
    const handleSearch = (e) => {
        e.preventDefault();
        console.log('sirvo');
        history.push(`?q=${search}`);
    };

    return (
        <div>
            <h1>Search Screen</h1>
            <hr />
            <div className="row">
                <div className="col-5">
                    <h4>Search Form</h4>
                    <hr/>
                    <form>
                        <input 
                            type="search"
                            placeholder="Find your hero"
                            autoComplete="off"
                            className="form-control"
                            name="search"
                            value={search}
                            onChange={handleInputChange}
                        />
                        <div className="m-1 d-grid gap-2">
                            <button 
                                type="submit"
                                className="btn btn-outline-primary"
                                onClick={handleSearch}
                            >
                                Search...
                            </button>
                        </div>
                        
                    </form>
                </div>
                <div className="col-7">
                    <h4>Results</h4>
                    <hr />
                    {
                        ((q==='') && 
                        <div className="alert alert-info">
                            Search a hero
                        </div>) ||
                        ((q!=='' && heroesFiltered.length === 0) && 
                        <div className="alert alert-warning">
                            There is no hero
                        </div>)
                    }
                    {
                        heroesFiltered.map( hero => (
                            <HeroCard key={hero.id} {...hero}/>
                        ))
                    }
                </div>

            </div>
        </div>
    );
};
