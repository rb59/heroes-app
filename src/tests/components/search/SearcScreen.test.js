import {mount} from 'enzyme';
import { MemoryRouter, Route } from 'react-router-dom';
import { SearchScreen } from '../../../components/search/SearchScreen';
// cd
describe('Pruebas en SearchScreen', () => {
    
    const history = {
        push: jest.fn(),
    };
    test('debe mostrarse correctamente con valores por defecto', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <Route 
                    path="/search" 
                    component={() => <SearchScreen /> } 
                />
            </MemoryRouter>
        );
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-info').text().trim()).toBe('Search a hero');
    });

    test('debe mostrar el heroe y el input con el valor del queryString', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route 
                    path="/search" 
                    component={() => <SearchScreen /> } 
                />
            </MemoryRouter>
        );
        expect(wrapper.find('input').prop('value')).toBe('batman');
        expect(wrapper).toMatchSnapshot();
    });

    test('debe mostrar un error si no se encuentra el hero', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman1234']}>
                <Route 
                    path="/search" 
                    component={() => <SearchScreen /> } 
                />
            </MemoryRouter>
        );
        expect(wrapper.find('.alert-warning').exists()).toBe(true);
        expect(wrapper.find('.alert-warning').text().trim()).toBe('There is no hero');
        expect(wrapper).toMatchSnapshot();
    });
    
    test('debe llamar el push del history', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <Route 
                    path="/search" 
                    component={() => <SearchScreen history={history} /> } 
                />
            </MemoryRouter>
        );
        wrapper.find('input').simulate('change',{
            target: {
                name: 'search',
                value: 'batman',
            },
        });
        const submit = wrapper.find('button').prop('onClick');
        // console.log(submit);
        submit({
            preventDefault(){},
        });
        expect(history.push).toHaveBeenCalledWith('?q=batman')
    });
    
});
