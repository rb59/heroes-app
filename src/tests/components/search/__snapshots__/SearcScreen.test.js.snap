// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en SearchScreen debe mostrar el heroe y el input con el valor del queryString 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form>
        <input
          autoComplete="off"
          className="form-control"
          name="search"
          onChange={[Function]}
          placeholder="Find your hero"
          type="search"
          value="batman"
        />
        <div
          className="m-1 d-grid gap-2"
        >
          <button
            className="btn btn-outline-primary"
            onClick={[Function]}
            type="submit"
          >
            Search...
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="card ms-3 mb-3"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row g-0"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Batman"
              className="card-img"
              src="./assets/heroes/dc-batman.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-tittle"
              >
                Batman
              </h5>
              <p
                className="card-text"
              >
                Bruce Wayne
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  Detective Comics #27
                </small>
              </p>
              <a
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                MÃ¡s...
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en SearchScreen debe mostrar un error si no se encuentra el hero 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form>
        <input
          autoComplete="off"
          className="form-control"
          name="search"
          onChange={[Function]}
          placeholder="Find your hero"
          type="search"
          value="batman1234"
        />
        <div
          className="m-1 d-grid gap-2"
        >
          <button
            className="btn btn-outline-primary"
            onClick={[Function]}
            type="submit"
          >
            Search...
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-warning"
      >
        There is no hero
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en SearchScreen debe mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form>
        <input
          autoComplete="off"
          className="form-control"
          name="search"
          onChange={[Function]}
          placeholder="Find your hero"
          type="search"
          value=""
        />
        <div
          className="m-1 d-grid gap-2"
        >
          <button
            className="btn btn-outline-primary"
            onClick={[Function]}
            type="submit"
          >
            Search...
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a hero
      </div>
    </div>
  </div>
</div>
`;
